backends/ignite.go

```diff
@@ -0,0 +1,256 @@
+package backends
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/prebid/prebid-cache/config"
+	"github.com/prebid/prebid-cache/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+// IgniteBackend implements Backend interface and communicates with the Apache Ignite storage
+// via its REST API as documented in https://ignite.apache.org/docs/2.11.1/restapi#rest-api-reference
+type IgniteBackend struct {
+	sender    requestSender
+	serverURL *url.URL
+	headers   http.Header
+	cacheName string
+}
+
+// httpClientWrapper lets us mock the http.Client
+type httpClientWrapper interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// requestSender defines a DoRequest method that will let us send the request to the Ignite server
+// and handle it's response and error. Other implementations of it will let us mock errorscenarios.
+type requestSender interface {
+	DoRequest(ctx context.Context, url *url.URL, headers http.Header) ([]byte, error)
+}
+
+// igniteSender implements the requestSender interface
+type igniteSender struct {
+	httpClient httpClientWrapper
+}
+
+// DoRequest will hit the Ignite server specified in the url parameter and handle error responses
+func (c *igniteSender) DoRequest(ctx context.Context, url *url.URL, headers http.Header) ([]byte, error) {
+	httpReq, err := http.NewRequestWithContext(ctx, "GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(headers) > 0 {
+		httpReq.Header = headers
+	}
+
+	httpResp, httpErr := c.httpClient.Do(httpReq)
+	if httpErr != nil {
+		return nil, httpErr
+	}
+
+	if httpResp.StatusCode != http.StatusOK {
+		httpErr = fmt.Errorf("Ignite error. Unexpected status code: %d", httpResp.StatusCode)
```

Comment:
If we return here, we'd get an entry in the logs with the status code alone, which is not that helpful. The logic behind not exiting here is to append more information to `httpErr` such as the error message found in `httpResp.Body` or lack thereof.

Reply:
Thoughts?

Reply:
I see. I can get on board with this approach, however, I'm curious if/when the additional details are needed. For example, typically if the status code is 204 that means that the request was fulfilled without a body so if that is a status code that Ignite might send back, I don't see why we might need to append "Received empty httpResp.Body" to the error message. Shouldn't we be able to just report back that status code 204 was received from Ignite and rely on that status code's meaning to tell the story?

I guess I'm asking if you've observed a need to provide this additional information when the status code is not 200 when working with a live Ignite server. In most cases the semantics of a HTTP code provide the information we need. Furthermore, I'm curious if you've observed any instances where Ignite provides an error message in the body along with the non-200 status code. If so, would that suffice?

Reply:
Discussed offline. The Ignite REST API docs imply that all responses are 200s and we have yet to encounter a non-200 response when working with a live Ignite server. Any error details are expected to be in the response body. As such, I'm good with the current design which is focused on receiving and unpacking a response while catching transport and response body errors but also accounting for the possibility of non-200 error codes. A non-200 error code is returned but only after we've attempted to read the response body since there might be more information there.